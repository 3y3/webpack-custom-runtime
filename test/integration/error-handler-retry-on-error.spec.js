'use strict';

const cct = require('./utils/custom-configurable-test');
const WebpackCustomRuntime = require('../../').CustomRuntimePlugin;

describe('RetryOnError', function() {

    const base = cct.base({
        plugins: [
            new WebpackCustomRuntime({ behavior: '' }),
            new WebpackCustomRuntime.ErrorHandler.RetryOnError()
        ]
    });

    function cit(name, plugins = [], base) {
        cct(name, {
            plugins: [
                new WebpackCustomRuntime({ behavior: '' }),
                ...plugins
            ]
        }, base);
    }

    cit('should setup default retry', [
        new WebpackCustomRuntime.ErrorHandler.RetryOnError()
    ]);

    cit('should configure retry', [
        new WebpackCustomRuntime.ErrorHandler.RetryOnError({
            maxRetryCount: 2
        })
    ], base);

    cit('should configure waitOnline', [
        new WebpackCustomRuntime.ErrorHandler.RetryOnError({
            waitOnline: true
        })
    ], base);

    cit('should be disabled on `enabled: false`', [
        new WebpackCustomRuntime.ErrorHandler.RetryOnError({
            enabled: false
        })
    ]);

    describe('should configure namespace', function() {

        cit('top level case', [
            new WebpackCustomRuntime.ErrorHandler.RetryOnError({
                namespace: '__NS__'
            })
        ], base);

        cit('brackets case', [
            new WebpackCustomRuntime.ErrorHandler.RetryOnError({
                namespace: 'window["__NS__"]'
            })
        ], base);

        cit('dot case', [
            new WebpackCustomRuntime.ErrorHandler.RetryOnError({
                namespace: 'Ya.__NS__'
            })
        ], base);

        cit('requireFn case', [
            new WebpackCustomRuntime.ErrorHandler.RetryOnError({
                namespace: '{requireFn}.__NS__'
            })
        ], base);
    });
});
