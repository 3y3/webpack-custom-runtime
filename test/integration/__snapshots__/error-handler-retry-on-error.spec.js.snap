// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RetryOnError should configure namespace brackets case: runtime.js 1`] = `
"Index: runtime.js
===================================================================
--- runtime.js	
+++ runtime.js	
@@ Change 1 @@
-/******/ 	      var ns = __webpack_require__.RetryOnError;
+/******/ 	      var ns = window['__NS__'];
@@ Change 2 @@
-/******/ 	            __webpack_require__.RetryOnError = __webpack_require__.RetryOnError || {};
+/******/ 	            window['__NS__'] = window['__NS__'] || {};
@@ Change 3 @@
-/******/ 	            var ns = __webpack_require__.RetryOnError;
+/******/ 	            var ns = window['__NS__'];
"
`;

exports[`RetryOnError should configure namespace dot case: runtime.js 1`] = `
"Index: runtime.js
===================================================================
--- runtime.js	
+++ runtime.js	
@@ Change 1 @@
-/******/ 	      var ns = __webpack_require__.RetryOnError;
+/******/ 	      var ns = Ya.__NS__;
@@ Change 2 @@
-/******/ 	            __webpack_require__.RetryOnError = __webpack_require__.RetryOnError || {};
+/******/ 	            Ya.__NS__ = Ya.__NS__ || {};
@@ Change 3 @@
-/******/ 	            var ns = __webpack_require__.RetryOnError;
+/******/ 	            var ns = Ya.__NS__;
"
`;

exports[`RetryOnError should configure namespace requireFn case: runtime.js 1`] = `
"Index: runtime.js
===================================================================
--- runtime.js	
+++ runtime.js	
@@ Change 1 @@
-/******/ 	      var ns = __webpack_require__.RetryOnError;
+/******/ 	      var ns = __webpack_require__.__NS__;
@@ Change 2 @@
-/******/ 	            __webpack_require__.RetryOnError = __webpack_require__.RetryOnError || {};
+/******/ 	            __webpack_require__.__NS__ = __webpack_require__.__NS__ || {};
@@ Change 3 @@
-/******/ 	            var ns = __webpack_require__.RetryOnError;
+/******/ 	            var ns = __webpack_require__.__NS__;
"
`;

exports[`RetryOnError should configure namespace top level case: runtime.js 1`] = `
"Index: runtime.js
===================================================================
--- runtime.js	
+++ runtime.js	
@@ Change 1 @@
-/******/ 	      var ns = __webpack_require__.RetryOnError;
+/******/ 	      var ns = window.__NS__;
@@ Change 2 @@
-/******/ 	            __webpack_require__.RetryOnError = __webpack_require__.RetryOnError || {};
+/******/ 	            window.__NS__ = window.__NS__ || {};
@@ Change 3 @@
-/******/ 	            var ns = __webpack_require__.RetryOnError;
+/******/ 	            var ns = window.__NS__;
"
`;

exports[`RetryOnError should configure retry: runtime.js 1`] = `
"Index: runtime.js
===================================================================
--- runtime.js	
+++ runtime.js	
@@ Change 1 @@
-/******/ 	            ns.max = ns.max || (1 || 0);
+/******/ 	            ns.max = ns.max || (2 || 0);
"
`;

exports[`RetryOnError should configure waitOnline: runtime.js 1`] = `
"Index: runtime.js
===================================================================
--- runtime.js	
+++ runtime.js	
@@ Change 1 @@
-/******/ 	      return Promise.resolve()
+/******/ 	      return ns
+/******/ 	        .waitOnline()
@@ Change 2 @@
+/******/ 	                ns.waitOnline = function() {
+/******/ 	                    return new Promise(function(resolve) {
+/******/ 	                        if (navigator.onLine) {
+/******/ 	                            resolve();
+/******/ 	                        } else {
+/******/ 	                            window.addEventListener('online', function onOnline() {
+/******/ 	                                window.removeEventListener('online', onOnline);
+/******/ 	                                resolve();
+/******/ 	                            });
+/******/ 	                        }
+/******/ 	                    });
+/******/ 	                };
+/******/ 	            
+/******/
"
`;

exports[`RetryOnError should setup default retry: runtime.js 1`] = `
"Index: runtime.js
===================================================================
--- runtime.js	
+++ runtime.js	
@@ Change 1 @@
-/******/ 	  return Promise.reject(originalError);
+/******/ 	  return [
+/******/ 	    function(installedChunks, chunkId, originalError, result) {
+/******/ 	      var ns = __webpack_require__.RetryOnError;
+/******/ 	      var retries = ns.retries[chunkId] || 0;
+/******/
+/******/ 	      if (retries >= ns.max) {
+/******/ 	        return result;
+/******/ 	      }
+/******/
+/******/ 	      ns.retries[chunkId] = retries + 1;
+/******/
+/******/ 	      return Promise.resolve()
+/******/ 	        .then(function() {
+/******/ 	          return __webpack_require__.e(chunkId);
+/******/ 	        })
+/******/ 	        .then(function() {}); // Resolve empty value
+/******/ 	    }
+/******/ 	  ]
+/******/ 	    .reduce(function(promise, strategy) {
+/******/ 	      return promise.then(function(result) {
+/******/ 	        return strategy(installedChunks, chunkId, originalError, result);
+/******/ 	      });
+/******/ 	    }, Promise.resolve(originalError))
+/******/ 	    .then(function(result) {
+/******/ 	      return result ? Promise.reject(result) : Promise.resolve();
+/******/ 	    });
@@ Change 2 @@
+/******/ 	            __webpack_require__.RetryOnError = __webpack_require__.RetryOnError || {};
+/******/
+/******/ 	            var ns = __webpack_require__.RetryOnError;
+/******/ 	            ns.max = ns.max || (1 || 0);
+/******/ 	            ns.retries = ns.retries || {};
+/******/ 	        
+/******/
"
`;
